pkg/core/klinedriver/klinebuilder.go

```diff
@@ -0,0 +1,295 @@
+			// no trade happened up to the updateTime
+			// use the last kline to fill the gap
+			accBucket.KLine.Open = lastBucket.KLine.Close
+			accBucket.KLine.High = lastBucket.KLine.Close
+			accBucket.KLine.Low = lastBucket.KLine.Close
+			accBucket.KLine.Close = lastBucket.KLine.Close
+			accBucket.KLine.LastTradeID = lastBucket.KLine.LastTradeID
+			accBucket.KLine.Volume = fixedpoint.Zero
+			accBucket.KLine.QuoteVolume = fixedpoint.Zero
+			accBucket.KLine.TakerBuyBaseAssetVolume = fixedpoint.Zero
+			accBucket.KLine.TakerBuyQuoteAssetVolume = fixedpoint.Zero
+		}
+		// two cases from here:
+		// - the update time is in the accumulating bucket
+		// - the update time is after the accumulating bucket
+		//   - there might be multiple buckets in between
+
+		// check if the update time is in the accumulating bucket
+		if accBucket.Contains(updateTime) {
+			// update time is in the accumulating bucket, just add kline
+			accBucket.KLine.EndTime = types.Time(updateTime.Time().UTC())
+			kline := *accBucket.KLine
+			kLinesMap[interval] = append(kLinesMap[interval], &kline)
+			continue
+		}
+
+		// update time is after the accumulating bucket
+		// we should close the current accumulating bucket and add kline
+		accBucket.KLine.Closed = true
+		accBucket.KLine.EndTime = types.Time(accBucket.EndTime.Time().UTC())
+		kline := *accBucket.KLine
+		kLinesMap[interval] = append(kLinesMap[interval], &kline)
+		accBucket.Exposed = true
+		// find the next containing bucket and reset the accumulating bucket
+		nextBucket, numShifts := accBucket.findNextBucket(updateTime)
+		nextBucket = kb.resetKLine(interval, nextBucket.StartTime)
+		// fill the gaps
+		filledKLines := accBucket.FillGapKLines(numShifts)
+		kLinesMap[interval] = append(kLinesMap[interval], filledKLines...)
+
+		nextBucket.KLine.Open = accBucket.KLine.Close
+		nextBucket.KLine.High = accBucket.KLine.Close
+		nextBucket.KLine.Low = accBucket.KLine.Close
+		nextBucket.KLine.Close = accBucket.KLine.Close
+		nextBucket.KLine.LastTradeID = accBucket.KLine.LastTradeID
+		kb.lastBucketMap[interval] = accBucket
+		// add an opening kline
+		nextBucket.KLine.EndTime = types.Time(updateTime.Time().UTC())
+		nextBucket.KLine.Closed = false
+		nextKLine := *nextBucket.KLine
+		kLinesMap[interval] = append(kLinesMap[interval], &nextKLine)
+	}
+	return kLinesMap
+}
+
+func (kb *KLineBuilder) resetKLine(resetInterval types.Interval, resetTime types.Time) *Bucket {
+	resetTimeUTC := resetTime.Time().UTC()
+	newBucket := &Bucket{
+		StartTime: resetTime,
+		EndTime:   types.Time(resetTime.Time().Add(resetInterval.Duration())),
+		Interval:  resetInterval,
+		KLine: &types.KLine{
+			Symbol:    kb.symbol,
+			Interval:  resetInterval,
+			StartTime: types.Time(resetTimeUTC),
+			EndTime:   types.Time(resetTimeUTC.Add(time.Millisecond)),
+			Closed:    false,
+		},
+	}
+	kb.accBucketMap[resetInterval] = newBucket
+	lastBucket, hasLast := kb.lastBucketMap[resetInterval]
```

Comment:
if lastBucket is only used in the clause, you can write:

```
if lastBucket, hasLast := kb.lastBucketMap[resetInterval]; hasLast {
  ... 
}
```

Reply:
Good point.
I didn't notice this as I write the code.